### PR TITLE
Fix: Ensure redirect to login when user is not logged and click "Enroll Button"

### LINKED ISSUE #1571: [Bug]: Better error message on tracks page
### What happened?

When not logged in and you visit a tracks page to enroll you get a generic error. It should at least tell you to log in to enroll or better maybe prompt them/redirect them to login and then do the action?

Showing a proper error message is fine for now though

### What browsers are you seeing the problem on?

_No response_

### Relevant log output

```zsh
visit https://typehero.dev/tracks/typescript-foundations while not logged in.

click enroll
```

### PR DESCRIPTION
**Pull request by 17AnuragMishra**
…rollment

## Description

**Fix:** Redirect users to the login page when attempting to enrol in a track while not logged in.

**Description:** 
This PR fixes the issue where an error message was shown every time a user tried to enrol in a track while not logged in. Instead of just showing an error, the user can log in with a proper redirect.

- Added a check for the `statusCode` and `message` in the error handler.
- A clear login prompt was provided in the toast notification.

## Related Issue

Closes #1571

## Motivation and Context

This change is required to fix the issue where users trying to enroll in a track while not logged in would always see a generic error message. Instead of a clear prompt asking the user to log in, the system was incorrectly showing a failure to enroll message.

The fix ensures that if the user is not logged in, they are prompted to log in via a toast notification, or redirected to the login page, solving the confusion and improving user experience.

Fixes #1571

### CODE REVIEW

File: apps/web/src/app/tracks/_components/enroll-button.tsx [hunk 1]
```diff
@@ -28,11 +33,12 @@ export function ActionButton({ action, text, trackId, slug }: EnrollButtonProps)
...
-        } catch (error) {
-          error instanceof Error && error.message === 'User is not logged in'
-            ? toast({
-                title: `Please Login to continue`,
+        } catch (error: unknown) { // Here we use `unknown` to dealwith unknow errors
+          if (error instanceof Error) {
+            const customError = error as CustomError;
+            if (customError.message === 'User is not logged in' || customError.statusCode === 401) {
+              toast({
+                title: 'Please Login to continue',
```
Comment (bautistaaa) [inline apps/web/src/app/tracks/_components/enroll-button.tsx:41]: Can we add a redirect back to the page they are on, I believe you can just add a queryparam. check the codebase for examples

File: apps/web/src/app/tracks/_components/enroll-button.tsx [hunk 2]
```diff
@@ -42,14 +48,25 @@ export function ActionButton({ action, text, trackId, slug }: EnrollButtonProps)
...
                   <p>There was an error trying to {text.toLowerCase()} you from the track.</p>
                 ),
               });
+            }
+          } else {
+            // Handle unexpected non-Error objects if needed
+            toast({
+              title: 'Unknown Error',
+              variant: 'destructive',
+              description: <p>Something went wrong, but the error type is not recognized.</p>,
```
[the remark this answers is not in the document]
Comment (17AnuragMishra) [reply apps/web/src/app/tracks/_components/enroll-button.tsx:66]: okay, got it! I'll made the required changes..

File: apps/web/src/app/tracks/_components/enroll-button.tsx (hunk 1 above)
Comment (17AnuragMishra) [reply apps/web/src/app/tracks/_components/enroll-button.tsx:41]: Okay, I'll do this as well.

**Comment by 17AnuragMishra** [comment]
Could you pls tell me why some checks got rejected.!

**Comment by bautistaaa** [comment]
> Could you pls tell me why some checks got rejected.!

you can click on them and read the error output

**Comment by 17AnuragMishra** [comment]
I have nothing to do with this right, sir.
Fixed the changes and now it will redirect the user to the login page as you asked me to implement.

**Comment by bautistaaa** [comment]
> image I have nothing to do with this right, sir. Fixed the changes and now it will redirect the user to the login page as you asked me to implement.

no, but run pnpm checks:fix cause your formatting is failing

**Comment by bautistaaa** [comment]
i just hit the same error you got `NO GITHUB ID` going to try and fix asap

**Comment by bautistaaa** [comment]
nm i accidently commented out something in my `.env` somewhow

**Comment by 17AnuragMishra** [comment]
Pls, check and review again! Thank you

**Comment by 17AnuragMishra** [comment]
@bautistaaa sir pls review this, as I think it's ready for merge.

**Comment by aakash19here** [comment]
oops  20k + lines of changes, ig prettier added single quotes everywhere , you'll need to reset some changes my man

**Comment by 17AnuragMishra** [comment]
@aakash19here ok, I'll just commit the main enrol button stuff, The rest of the part going to be ignored now, but will it work?

**Comment by aakash19here** [comment]
you'll need to run `checks:fix` to fix the formatting

**Comment by 17AnuragMishra** [comment]
I'm going to revise or delete this repository and create a new pull request so that it doesn't become messy.

**Comment by bautistaaa** [comment]
going to close this and reopen when you have your changes in order

**Pull request closed without merging**